examples/sponsored-tx-enabled-auction/src/lib.rs

```diff
@@ -0,0 +1,432 @@
+)]
+fn auction_bid(ctx: &ReceiveContext, host: &mut Host<State>) -> ReceiveResult<()> {
+    // Ensure the sender is the cis2 token contract.
+    if !ctx.sender().matches_contract(&host.state().cis2_contract) {
+        bail!(Error::NotTokenContract.into());
+    }
+
+    // Getting input parameters.
+    let params: OnReceivingCis2Params<ContractTokenId, ContractTokenAmount> =
+        ctx.parameter_cursor().get()?;
+
+    // Ensure that only accounts can bid for an item.
+    let bidder_address = match params.from {
+        Address::Contract(_) => bail!(Error::OnlyAccount.into()),
+        Address::Account(account_address) => account_address,
+    };
+
+    // Get the item_index from the additionalData.
+    let additional_data_index: AdditionalDataIndex = from_bytes(params.data.as_ref())?;
+
+    let mut item = host
+        .state_mut()
+        .items
+        .entry(additional_data_index.item_index)
+        .occupied_or(Error::NoItem)?;
+
+    // Ensure the token_id matches.
+    ensure_eq!(item.token_id, params.token_id, Error::WrongTokenID.into());
+
+    // Ensure the auction has not been finalized yet.
+    ensure_eq!(item.auction_state, AuctionState::NotSoldYet, Error::AuctionAlreadyFinalized.into());
+
+    let slot_time = ctx.metadata().slot_time();
+    // Ensure the auction has not ended yet.
+    ensure!(slot_time <= item.end, Error::BidTooLate.into());
+
+    // Ensure that the new bid exceeds the highest bid so far.
+    let old_highest_bid = item.highest_bid;
+    ensure!(params.amount > old_highest_bid, Error::BidNotGreaterCurrentBid.into());
+
+    // Set the new highest_bid.
+    item.highest_bid = params.amount;
+
+    if let Some(account_address) = item.highest_bidder.replace(bidder_address) {
+        // Refunding old highest bidder.
+        // This transfer (given enough NRG of course) always succeeds because
+        // the `account_address` exists since it was recorded when it
+        // placed a bid. If an `account_address` exists, and the
+        // contract has the funds then the transfer will always succeed.
+        // Please consider using a pull-over-push pattern when expanding this
+        // smart contract to allow smart contract instances to
+        // participate in the auction as well. https://consensys.github.io/smart-contract-best-practices/attacks/denial-of-service/
+        let parameter = TransferParameter {
+            0: vec![Transfer {
+                token_id: item.token_id,
+                amount:   old_highest_bid,
+                from:     concordium_std::Address::Contract(ctx.self_address()),
+                to:       concordium_cis2::Receiver::Account(account_address),
+                data:     AdditionalData::empty(),
+            }],
+        };
+
+        drop(item);
+
+        host.invoke_contract(
+            &host.state().cis2_contract.clone(),
+            &parameter,
+            EntrypointName::new_unchecked("transfer"),
+            Amount::zero(),
+        )?;
```

Comment:
**Question:** Did you consider using the [`concordium_cis2::Cis2Client`](https://docs.rs/concordium-cis2/latest/concordium_cis2/struct.Cis2Client.html)?